File: crud-item-list/src/main/java/com/gurunars/crud_item_list/ActionMoveDown.java
package com.gurunars.crud_item_list;

import java.util.List;
import java.util.Set;

import java8.util.function.BiFunction;
import java8.util.function.Function;

final class ActionMoveDown<ItemType> implements Action<ItemType> {

    private final BiFunction<List<ItemType>, Set<ItemType>, List<Integer>> positionFetcher =
            new PositionFetcher<>();

    private final Function<List<Integer>, Boolean> solidChunkChecker =
            new CheckerSolidChunk();

    @Override
    public boolean perform(final List<ItemType> all, final Set<ItemType> selectedItems) {
        List<Integer> positions = positionFetcher.apply(all, selectedItems);
        int positionToMoveUp = positions.get(positions.size()-1) + 1;
        ItemType itemToMoveUp = all.get(positionToMoveUp);
        all.remove(positionToMoveUp);
        all.add(positions.get(0), itemToMoveUp);
        return true;
    }

    @Override
    public boolean canPerform(List<ItemType> all, Set<ItemType> selectedItems) {
        List<Integer> positions = positionFetcher.apply(all, selectedItems);
        return solidChunkChecker.apply(positions) && positions.get(positions.size() - 1) < all.size() - 1;
    }

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/ActionMoveDown.kt
package com.gurunars.crud_item_list

import com.gurunars.box.IRoBox
import com.gurunars.box.merge
import com.gurunars.box.oneWayBranch
import com.gurunars.item_list.Item

internal class ActionMoveDown<ItemType : Item> : Action<ItemType> {

    override fun canPerform(
        all: IRoBox<List<ItemType>>,
        selectedItems: IRoBox<Set<ItemType>>
    ): IRoBox<Boolean> = merge(all, selectedItems).oneWayBranch {
        val positions = getPositions(first, second)
        isSolidChunk(positions) && positions[positions.size - 1] < first.size - 1
    }

    override fun perform(
        all: List<ItemType>,
        selectedItems: Set<ItemType>,
        consumer: ItemSetChange<ItemType>
    ) {
        val positions = getPositions(all, selectedItems)
        val positionToMoveUp = positions[positions.size - 1] + 1
        val itemToMoveUp = all[positionToMoveUp]
        consumer(mutableListOf<ItemType>().apply {
            addAll(all)
            removeAt(positionToMoveUp)
            add(positions[0], itemToMoveUp)
        }, selectedItems)
    }

}
